sys/syz-extract/netbsd.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2017 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/google/syzkaller/pkg/compiler"
+)
+
+type netbsd struct{}
+
+func (*netbsd) prepare(sourcedir string, build bool, arches []string) error {
+	if sourcedir == "" {
+		return fmt.Errorf("provide path to kernel checkout via -sourcedir flag (or make extract SOURCEDIR)")
+	}
+	if !build {
+		return fmt.Errorf("netbsd requires -build flag")
+	}
+	return nil
+}
+
+func (*netbsd) prepareArch(arch *Arch) error {
+	if err := os.Symlink(filepath.Join(arch.sourceDir, "sys", "arch", "amd64", "include"),
+		filepath.Join(arch.buildDir, "machine")); err != nil {
+		return fmt.Errorf("failed to create link: %v", err)
+	}
+	if err := os.Symlink(filepath.Join(arch.sourceDir, "sys", "arch", "x86", "include"),
+		filepath.Join(arch.buildDir, "x86")); err != nil {
+		return fmt.Errorf("failed to create link: %v", err)
+	}
+	return nil
+}
+
+func (*netbsd) processFile(arch *Arch, info *compiler.ConstInfo) (map[string]uint64, map[string]bool, error) {
+	args := []string{
+		"-fmessage-length=0",
+		"-nostdinc",
+		"-D_KERNEL",
```

Comment:
Why `_KERNEL`? Nothing in userland should define it.

Reply:
We are extracting constants from kernel sources. So in some sense we are kernel code, not user-space. You consider this as we trying to build a mini-kernel that just prints values of some consts.
But I don't know if it's required for netbsd or not (was required for linux and freebsd).
I would suggest to remove (comment out) all of these defines (including __BSD_VISIBLE/_NETBSD_SOURCE/etc) and then try to execute extract and add macros back as needed.

Reply:
If you are `_KERNEL` code, you are supposed to use headers in `/usr/src/sys/sys` or similar path where the kernel is located (`src/sys/`). Some headers are shared with userspace and libstand (for bootloader etc), but there are still kernel-only parts and we don't install everything from the kernel namespace into `/usr/include/`. This means that definition of `_KERNEL` and including NetBSD headers will lead to unresolvable missing symbols.

There is a special define to read kernel data structures in certain userland programs: `_KMEMUSER`. And certain programs of the similar type need `_KERNTYPES`.

Reply:
As far as I see, we do use these headers. See -I below, where sourceDir is meant to be netbsd kernel checkout.

Reply:
Just for the curiosity, there is an example of a kernel module. S stands for path to the kernel sources.

https://github.com/NetBSD/src/blob/trunk/sys/modules/examples/ping/Makefile#L5

Reply:
I see! So, I hope it will work!

Reply:
This is not a module. This builds a normal user-space program with main.

Reply:
Right, I was sharing an interesting detail.